In [79]:
#Text sentiment analysis:

#importing libraries:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re # for finding or removing unecessary words:
import string # for puncation and string constants: 
import nltk # Natural langauage toolkit:
from nltk.corpus import stopwords # common words (is,like,the,and) removed for better focus:
from nltk.stem import WordNetLemmatizer # reduce the word to its short form (i.e running to run)

# neccesary downloads:
nltk.download("stopwords")
nltk.download('punkt')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Arhum\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Arhum\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Arhum\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [80]:
# for feature extraction:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score

In [81]:
df = pd.read_csv("E:/University Work/Semester_4/Valid.csv")
df

,text,label
0,It's been about 14 years since Sharon Stone aw...,0
1,someone needed to make a car payment... this i...,0
2,The Guidelines state that a comment must conta...,0
3,This movie is a muddled mish-mash of clichés f...,0
4,Before Stan Laurel became the smaller half of ...,0
...,...,...
4995,"Man, I loved this movie! This really takes me ...",1
4996,Recovery is an incredibly moving piece of work...,1
4997,"You can take the crook out of the joint, but i...",1
4998,FUTZ is the only show preserved from the exper...,1


In [82]:
df.head()

,text,label
0,It's been about 14 years since Sharon Stone aw...,0
1,someone needed to make a car payment... this i...,0
2,The Guidelines state that a comment must conta...,0
3,This movie is a muddled mish-mash of clichés f...,0
4,Before Stan Laurel became the smaller half of ...,0


In [83]:
stop_words = set(stopwords.words('english'))
le = WordNetLemmatizer()
def preprocess(text):
    text = text.lower()
    text = re.sub(r'\d+','',text)
    text = re.sub(f'[{re.escape(string.punctuation)}]','',text)
    tokens = nltk.word_tokenize(text)
    tokens = [le.lemmatize(text) for word in tokens if word not in stop_words]
    return''.join(tokens)
df['clean_text'] = df['text'].apply(preprocess)

In [84]:
# Clean_text (free from punctuation marks, lemmataion)
df['clean_text']

0       its been about  years since sharon stone award...
1       someone needed to make a car payment this is t...
2       the guidelines state that a comment must conta...
3       this movie is a muddled mishmash of clichés fr...
4       before stan laurel became the smaller half of ...
                              ...                        
4995    man i loved this movie this really takes me ba...
4996    recovery is an incredibly moving piece of work...
4997    you can take the crook out of the joint but it...
4998    futz is the only show preserved from the exper...
4999    the mother tells of a recently widowed mids mo...
Name: clean_text, Length: 5000, dtype: object

In [85]:
# vectorizer:
# feature engineering using TF-IDF
vectorizer = TfidfVectorizer(max_features=5000)
X = vectorizer.fit_transform(df['clean_text'])# converts to numerical format.
# Encode 'sentiment' (assumes positive and negative labels)
y = df['label']

In [87]:
print(y)
y.isnull

0       0
1       0
2       0
3       0
4       0
       ..
4995    1
4996    1
4997    1
4998    1
4999    1
Name: label, Length: 5000, dtype: int64


<bound method Series.isnull of 0       0
1       0
2       0
3       0
4       0
       ..
4995    1
4996    1
4997    1
4998    1
4999    1
Name: label, Length: 5000, dtype: int64>

In [78]:
df

,text,label,clean_text
0,It's been about 14 years since Sharon Stone aw...,0,its been about years since sharon stone award...
1,someone needed to make a car payment... this i...,0,someone needed to make a car payment this is t...
2,The Guidelines state that a comment must conta...,0,the guidelines state that a comment must conta...
3,This movie is a muddled mish-mash of clichés f...,0,this movie is a muddled mishmash of clichés fr...
4,Before Stan Laurel became the smaller half of ...,0,before stan laurel became the smaller half of ...
...,...,...,...
4995,"Man, I loved this movie! This really takes me ...",1,man i loved this movie this really takes me ba...
4996,Recovery is an incredibly moving piece of work...,1,recovery is an incredibly moving piece of work...
4997,"You can take the crook out of the joint, but i...",1,you can take the crook out of the joint but it...
4998,FUTZ is the only show preserved from the exper...,1,futz is the only show preserved from the exper...


In [89]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state = 42)

In [91]:
# ML - MODEL:

lr = LogisticRegression()
lr

LogisticRegression()

In [92]:
lr.fit(X_train,y_train)

LogisticRegression()

In [93]:
y_pred = lr.predict(X_test)
y_pred

array([1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0,
       0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0,
       0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0,
       0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1,
       1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1,
       1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0,
       0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0,

In [95]:
print("\n-------Report----------")
print(f"\nAccuracy of sentiment_analysis {accuracy_score(y_test,y_pred)}")
print(f"\nClassification_report: {classification_report(y_test,y_pred)}")


-------Report----------

Accuracy of sentiment_analysis 0.86

Classification_report:               precision    recall  f1-score   support

           0       0.87      0.85      0.86       511
           1       0.85      0.87      0.86       489

    accuracy                           0.86      1000
   macro avg       0.86      0.86      0.86      1000
weighted avg       0.86      0.86      0.86      1000



In [100]:
conf_matrix = confusion_matrix(y_pred,y_test)
conf_matrix

array([[434,  63],
       [ 77, 426]], dtype=int64)

In [101]:
def predict_sentiment(text):
    cleaned = preprocess(text)
    vector = vectorizer.transform([cleaned])
    pred = lr.predict(vector)[0]
    return "Positive" if pred == 1 else "Negative"

print(predict_sentiment("This is absolutely amazing!"))


Positive


The model achieved an accuracy of 86% on the test set, which indicates solid performance.

📈 Classification Report:

Label	Precision	Recall	F1-Score	Support
Negative (0)	0.87	0.85	0.86	511
Positive (1)	0.85	0.87	0.86	489
Overall	0.86	0.86	0.86	1000

Precision: High precision for both classes shows that the model makes few false positive errors.

Recall: Balanced recall values indicate that the model is not biased towards any class.

F1-Score: F1-scores are equally strong, reinforcing the model’s balanced performance.

Correct Predictions: 434 (Negative) + 426 (Positive) = 860

Misclassifications: 63 (False Positives) + 77 (False Negatives)